python/lsst/daf/butler/_butler.py

```diff
@@ -230,8 +230,10 @@ def __init__(
             self._config: ButlerConfig = butler._config
         else:
             # Can only look for strings in the known repos list.
-            if isinstance(config, str) and config in self.get_known_repos():
-                config = str(self.get_repo_uri(config))
+            if isinstance(config, str):
+                # Somehow ButlerConfig fails in some cases if config is a
```

Comment:
Hmm. I wonder what that's about. I clearly knew at some point given I did force it back to str previously.

Reply:
If i strip `str()` then one unit test fails with this exception:
```
ERROR: testConstructor (__main__.S3DatastoreButlerTestCase)
Independent test of constructor.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/salnikov/lsst/gen3/daf_butler/tests/test_butler.py", line 548, in testConstructor
    butler = Butler(config_dir, run=self.default_run)
  File "/home/salnikov/lsst/gen3/daf_butler/python/lsst/daf/butler/_butler.py", line 238, in __init__
    self._config = ButlerConfig(config, searchPaths=searchPaths)
  File "/home/salnikov/lsst/gen3/daf_butler/python/lsst/daf/butler/_butlerConfig.py", line 95, in __init__
    uri = ResourcePath(other, forceDirectory=True)
  File "/u1/lsst/lsst_stack-6.0.0-new/conda/envs/lsst-scipipe-6.0.0/share/eups/Linux64/resources/geaf6fda20e+37bc81c514/python/lsst/resources/_resourcePath.py", line 214, in __new__
    raise RuntimeError(
RuntimeError: s3://anybucketname/6A6JK5LH67CLP5WFLC3U is already a file-like ResourcePath; cannot force it to directory.
```

Reply:
Ah. Right. The old problem with an arbitrary S3 URI not being able to tell the difference between being a file or a directory. Whereas downstream the defaulting code can say "does not end in butler.yaml therefore it must be a directory" and force it to be a directory -- which it can't easily do if it's already a ResourcePath which doesn't look like a directory (although maybe I could add some logic for that to work).